libraries/chain/controller.cpp

```diff
@@ -1130,9 +1130,19 @@ struct controller_impl {
       });
    }
 
-   bool fork_db_validated_block_exists( const block_id_type& id ) const {
+   bool fork_db_validated_block_exists( const block_id_type& id ) {
+       return fork_db.apply<bool>([&](const auto& forkdb) {
+         auto bsp = forkdb.get_block(id);
+         return bsp ? bsp->is_valid() : false;
```

Comment:
Might be good to add a comment that `is_valid()` is not thread safe for `block_state_legacy`.

Reply:
It is thread safe after https://github.com/AntelopeIO/spring/pull/788/commits/3c417b1f768d390299500a9033f200a3e8ecffaf

Reply:
We could make `is_valid()` public now and remove many of the `friends`.

Reply:
Unfortunately I could only remove:
- `template <typename BS> friend struct fork_database_impl;`
- `friend struct block_state_legacy_accessor;`

https://github.com/AntelopeIO/spring/pull/788/commits/311c94dfb0f29d32168111dd88f97918ff760b02